MicroK8s Day 2 scenario - GitOps with Helm (#141)
microsoft/azure_arc: azure_arc_k8s_jumpstart/docs/local_microk8s_gitops_helm.md

For example, an application (captured in an Helm chart) dev release can have no pod replication (single pod) while a production release, using the same chart can have 3 pod replicas. 

In the next section will use the "Hello Arc" Helm chart to deploy a production release which we will then change and see the results in real-time.

## Deployment Flow

For our scenario, we will deploy the "Hello Arc" application from the ["demo repository"](https://github.com/likamrat/hello_arc) through GitOps. We will deploy the "Hello Arc" application (a Namespace-level component) with 1 replica to the *prod* namespace.

![](../img/local_microk8s_gitops_helm/08.png)

![](../img/local_microk8s_gitops_helm/09.png)

## Deployment

To create the GitOps configuration and it's respective Kubernetes resources, we've provided a script in a shell file (Linux, MacOS) and batch file (Windows).

* Linux and MacOS

>     Edit the environment variables in the [*az_k8sconfig_helm_microk8s*](../microk8s/gitops/helm/az_k8sconfig_helm_microk8s.sh) shell script to match your parameters, and run it using the ```. ./az_k8sconfig_helm_microk8s.sh``` command.
> 

Reviewer: ```suggestion
    Edit the environment variables in the [*az_k8sconfig_helm_microk8s*](../microk8s/gitops/helm/az_k8sconfig_helm_microk8s.sh) shell script to match your parameters, and run the below command.
    ```console
    . ./az_k8sconfig_helm_microk8s.sh
    ``` 
```

    **Note**: The extra dot is due to the script having an *export* function and that needs to have the vars exported in the same shell session as the rest of the commands. 

* Windows

>     Edit the environment variables in the [*az_k8sconfig_helm_microk8s_windows*](../microk8s/gitops/helm/az_k8sconfig_helm_microk8s_windows.bat) batch file to match your parameters, and run it using the ```.\az_k8sconfig_helm_microk8s.bat``` command.

Reviewer: Change to a PowerShell script

## NGINX Controller Deployment
* Run the following command to install the nginx ingress controller on MicroK8s:
    ```shell
    kubectl apply -f https://raw.githubusercontent.com/kubernetes/ingress-nginx/controller-v0.35.0/deploy/static/provider/baremetal/deploy.yaml
    ```
* This command will create a new namespace and deploy the required components in this namespace. To verify the deployment of the ingress controller was successful, make sure the pod with name `ingress-nginx-controller-<random id>-<random id>` is in a running state with 1/1 containers ready and that a service has been exposed as NodePort.

    ```shell
    kubectl get pods -n ingress-nginx
    ```
    ![](../img/local_microk8s_gitops_helm/03.png)

    ```shell
    kubectl get svc -n ingress-nginx
    ```

    ![](../img/local_microk8s_gitops_helm/04.png)

* Take note of the port where the ingress has been exposed (in the image above it was assigned port **32106**). We now need to get the IP address assigned to our microk8s-vm instance in multipass:

    ```shell
    multipass list
    ```

    ![](../img/local_microk8s_gitops_helm/05.png)

> * Combining the IP address from multipass and the NodePort assigned to the ingress controller, we can now test that the NGINX ingress controller has been deployed successfully. In our case, the full address becomes http://172.22.206.155:32106, so if you open that in a browser or command line, you should get a HTTP 404 reponse with a nginx footer. This shows that the ingress is working. The 404 response is to be expected since you haven't setup an ingress route yet. You will do that in the next section.

Reviewer: ```suggestion
* Combining the IP address from multipass and the NodePort assigned to the ingress controller, we can now test that the NGINX ingress controller has been deployed successfully. In our case, the full address becomes *`http://172.22.206.155:32106`*. 

Using the below in your browser or command line, should get you with a HTTP 404 response with a nginx footer. This shows that the ingress is working. The 404 response is to be expected since you haven't setup an ingress route yet. You will do that in the next section.
```
